### PR TITLE
testutil: add MockCommand for testing code that runs programs

### PR DESCRIPTION
**Pull request by zyga**
This patch adds utilities for testing code that runs executables looked up on PATH.

The entry point is the ``MockCommand()`` function, taking name of a command an the exit code,
returning a ``* MockCmd`` object. The executable can be called normally on UNIX-like systems.
Support for PE-executables is not provided.

The list of calls made can be done with ``MockCmd.Calls()``.
To clean up after testing use defer call to ``MockCmd.Restore()``.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

**Comment by niemeyer** [comment]
LGTM. Please just test the error and panic if WriteFile returns something non-nil.

**Comment by zyga** [comment]
@elopio @fgimenez ^^ FYI

File: testutil/exec.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	logFile := path.Join(binDir, basename+".log")
+	countFile := path.Join(binDir, basename+".cnt")
+	err := ioutil.WriteFile(exeFile, []byte(fmt.Sprintf(""+
+		"#!/bin/sh\n"+
+		"echo \"$@\" >> %q\n"+
+		"exit %d\n", logFile, status)), 0700)
+	if err != nil {
+		panic(err)
+	}
+	os.Setenv("PATH", binDir+":"+os.Getenv("PATH"))
```
Comment (mvo5) [inline testutil/exec.go]: I think I would slightly prefer `os.Setenv("PATH", fmt.Sprintf("%s:%s", binDir, os.Getenv("PATH")))` here but thats probably personal preference (the + and the " and the : are hard to read(er) to read for me). But super nitpick, sorry.

**Comment by fgimenez** [comment]
Very cool! We have implemented the technique described at [1] to check other fields of `exec.Cmd`, not only `Args` (for instance, `Dir` or `Env`). Take a look at [2].

Thanks :)

[1] https://npf.io/2015/06/testing-exec-command/
[2] https://github.com/ubuntu-core/snappy/blob/master/integration-tests/testutils/cli/cli_test.go

**Pull request merged**